packages/swap/src/routes/quoteV2.ts

```diff
@@ -0,0 +1,365 @@
+import logger from '../utils/logger';
+import { getPools } from '../utils/pools';
+import {
+  getMinimumEgressAmount,
+  getEgressFee,
+  getIngressFee,
+  validateSwapAmount,
+} from '../utils/rpc';
+import ServiceError from '../utils/ServiceError';
+// import { getBrokerQuote } from '../utils/statechain';
+
+const fallbackChains = {
+  [Assets.ETH]: Chains.Ethereum,
+  [Assets.USDC]: Chains.Ethereum,
+  [Assets.FLIP]: Chains.Ethereum,
+  [Assets.BTC]: Chains.Bitcoin,
+  [Assets.DOT]: Chains.Polkadot,
+  [Assets.USDT]: Chains.Ethereum,
+} satisfies Record<Asset, Chain>;
+
+const quote = (io: Server) => {
+  const { handler, quotes$ } = getConnectionHandler();
+
+  io.on('connection', handler);
+
+  const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
+    const price = await getAssetPrice(asset);
+
+    if (typeof price !== 'number') return null;
+
+    return BigInt(new BigNumber(amount).times(price).toFixed(0));
+  };
+
+  async function quoteLeg(
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: null,
+  ): Promise<null>;
+  async function quoteLeg(
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: bigint,
+  ): Promise<bigint>;
+  async function quoteLeg(
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: bigint | null,
+  ): Promise<bigint | null>;
+  async function quoteLeg(
+    [baseAsset, quoteAsset]: readonly [InternalAsset, InternalAsset],
+    amount: bigint | null,
+  ): Promise<bigint | null> {
+    if (!amount) return null;
+
+    const id = randomUUID();
+
+    io.emit('quote_request', {
+      request_id: id,
+      amount: String(amount),
+      base_asset: getAssetAndChain(baseAsset),
+      quote_asset: getAssetAndChain(quoteAsset),
+    });
+
+    const quotes = await collectMakerQuotes(id, io.sockets.sockets.size, quotes$);
+
+    const { swappedAmount, remainingAmount } = await findPrice({
+      poolFee: 0,
+      amount: BigInt(amount),
+      limitOrders: quotes.flatMap(({ limit_orders: limitOrders }) =>
+        limitOrders.map(([tick, amt]) => ({ tick, amount: amt })),
+      ),
+    });
+
+    if (remainingAmount !== 0n) {
+      // TODO: handle remaining amount with the pool
+    }
+
+    return swappedAmount;
+  }
+
+  const getBestQuote = async (quoteRequest: ParsedQuoteParams & { amount: bigint }) => {
+    const { srcAsset, destAsset } = getInternalAssets(quoteRequest);
+    const networkFee = getPoolsNetworkFeeHundredthPips(env.CHAINFLIP_NETWORK);
+
+    const pools = await getPools(srcAsset, destAsset);
+
+    let intermediateAmount: bigint | null = null;
+    let outputAmount: bigint;
+    let networkFeeUsdc: bigint | null = null;
+    const firstLegPoolFee = getHundredthPipAmountFromAmount(
+      quoteRequest.amount,
+      pools[0].liquidityFeeHundredthPips,
+    );
+
+    const swapInputAmount = quoteRequest.amount - firstLegPoolFee;
```

Comment:
i could imagine that the code in the native addon already accounts for subtracting this from the input amount, but i havent looked into it

Reply:
it can, but i set the pool fee to zero so we can handle it here explicitly